libs/cmdio/render.go

```diff
@@ -46,8 +50,126 @@ func Heredoc(tmpl string) (trimmed string) {
 	return strings.TrimSpace(trimmed)
 }
 
-func renderJson(w io.Writer, v any) error {
-	pretty, err := fancyJSON(v)
+// writeFlusher represents a buffered writer that can be flushed. This is useful when
+// buffering writing a large number of resources (such as during a list API).
+type writeFlusher interface {
+	io.Writer
+	Flush() error
+}
+
+type jsonRenderer interface {
+	// Render an object as JSON to the provided writeFlusher.
+	renderJson(context.Context, writeFlusher) error
+}
+
+type textRenderer interface {
+	// Render an object as text to the provided writeFlusher.
+	renderText(context.Context, writeFlusher) error
+}
+
+type templateRenderer interface {
+	// Render an object using the provided template and write to the provided tabwriter.Writer.
+	renderTemplate(context.Context, *template.Template, *tabwriter.Writer) error
+}
+
+type readerRenderer struct {
+	reader io.Reader
+}
+
+func (r readerRenderer) renderText(_ context.Context, w writeFlusher) error {
+	_, err := io.Copy(w, r.reader)
+	if err != nil {
+		return err
+	}
+	return w.Flush()
+}
+
+type iteratorRenderer struct {
+	t          reflectIterator
+	bufferSize int
+}
+
+func (ir iteratorRenderer) getBufferSize() int {
+	if ir.bufferSize == 0 {
+		return 100
+	}
+	return ir.bufferSize
+}
+
+func (ir iteratorRenderer) renderJson(ctx context.Context, w writeFlusher) error {
+	// Iterators are always rendered as a list of resources in JSON.
+	_, err := w.Write([]byte("[\n  "))
```

Comment:
It is unfortunate that we didn't start out returning an object per entry, as it allows for streaming into `jq` or whatever other downstream tool. We might consider an option to enable this and that would affect this renderer. No action needed, observation only.